kafka-rest/src/main/java/io/confluent/kafkarest/ProducerMetrics.java

```diff
@@ -0,0 +1,395 @@
+      return tags;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      BeanCoordinate that = (BeanCoordinate) o;
+      return Objects.equals(beanName, that.beanName) && Objects.equals(tags, that.tags);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(beanName, tags);
+    }
+  }
+
+  /**
+   * A class representing a JMX MBean where each metric maps to an MBean attribute. Sensors should
+   * be added via the {@code sensor} methods on this class, rather than directly through the {@link
+   * Metrics} class, so that the sensor names are made to be unique (based on the MBean name) and so
+   * the sensors are removed when this group is {@link #close() closed}.
+   */
+  public class ProduceMetricMBean implements AutoCloseable {
+    private final BeanCoordinate beanCoordinate;
+
+    /**
+     * Create a produce metrics MBean.
+     *
+     * @param beanCoordinate the identifier of the bean; may not be null and must be valid
+     */
+    ProduceMetricMBean(BeanCoordinate beanCoordinate) {
+      Objects.requireNonNull(beanCoordinate);
+      this.beanCoordinate = beanCoordinate;
+    }
+
+    public void recordMetrics(String sensorName, double value) {
+      Sensor sensor =
+          metrics.getSensor(
+              String.join(":", JMX_PREFIX, ProducerMetricsRegistry.GROUP_NAME, sensorName));
+      if (sensor != null) {
+        sensor.record(value);
+      }
+    }
+
+    /**
+     * The {@link Metrics} that this MBean belongs to. Do not use this to add {@link Sensor
+     * Sensors}, since they will not be removed when this group is{@link #close() closed}. Metrics
+     * can be added directly, as long as the metric names are obtained from this group via the
+     * {@link #metricName(MetricNameTemplate)} method.
+     *
+     * @return the metrics; never null
+     */
+    public Metrics metrics() {
+      return metrics;
+    }
+
+    /**
+     * Create the name of a metric that belongs to this group and has the group's tags.
+     *
+     * @param template the name template for the metric; may not be null
+     * @return the metric name; never null
+     * @throws IllegalArgumentException if the name is not valid
+     */
+    public MetricName metricName(MetricNameTemplate template) {
+      return metrics.metricInstance(template, beanCoordinate.tags);
+    }
+
+    /**
+     * Get or create a sensor with the given unique name and no parent sensors. This uses a default
+     * recording level of INFO.
+     *
+     * @param name The sensor name
+     * @return The sensor
+     */
+    synchronized Sensor sensor(String name) {
+      Sensor sensor = metrics.sensor(name);
+      return sensor;
+    }
+
+    /** Remove all sensors and metrics associated with this group. */
+    @Override
+    public synchronized void close() {
```

Comment:
I'm not sure what are we trying to achieve with the mutual exclusion here. In case it's mostly avoiding an internal problem similar to `ConcurrentModificationException` I guess it's OK, but if not, I think we should clarify the lifecycle of this and enforce it with a simple state machine.